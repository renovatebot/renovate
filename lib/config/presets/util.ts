import { logger } from '../../logger';
import { ensureTrailingSlash } from '../../util/url';
import type { Preset } from './types';

export const PRESET_DEP_NOT_FOUND = 'dep not found';
export const PRESET_NOT_FOUND = 'preset not found';

export type PresetFetcher = (
  repo: string,
  fileName: string,
  endpoint: string
) => Promise<Preset>;

export type FetchPresetConfig = {
  pkgName: string;
  filePreset: string;
  presetPath?: string;
  endpoint: string;
  fetch: PresetFetcher;
};

export async function fetchPreset({
  pkgName,
  filePreset,
  presetPath,
  endpoint,
  fetch,
}: FetchPresetConfig): Promise<Preset | undefined> {
  // eslint-disable-next-line no-param-reassign
  endpoint = ensureTrailingSlash(endpoint);
  const [fileName, presetName, subPresetName] = filePreset.split('/');
  const pathPrefix = presetPath ? `${presetPath}/` : '';
  const buildFilePath = (name: string): string => `${pathPrefix}${name}`;
  let jsonContent: any | undefined;
  if (fileName === 'default') {
    try {
      jsonContent = await fetch(
        pkgName,
        buildFilePath('default.json'),
        endpoint
      );
    } catch (err) {
      if (err.message !== PRESET_DEP_NOT_FOUND) {
        throw err;
      }
      logger.debug('default.json preset not found - trying renovate.json');
      jsonContent = await fetch(
        pkgName,
        buildFilePath('renovate.json'),
        endpoint
      );
    }
  } else {
    jsonContent = await fetch(
      pkgName,
      buildFilePath(`${fileName}.json`),
      endpoint
    );
  }

  if (!jsonContent) {
    throw new Error(PRESET_DEP_NOT_FOUND);
  }
  if (presetName) {
    const preset = jsonContent[presetName];
    if (!preset) {
      throw new Error(PRESET_NOT_FOUND);
    }
    if (subPresetName) {
      const subPreset = preset[subPresetName];
      if (!subPreset) {
        throw new Error(PRESET_NOT_FOUND);
      }
      return subPreset;
    }
    return preset;
  }
  return jsonContent;
}
