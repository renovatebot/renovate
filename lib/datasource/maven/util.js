const got = require('../../util/got');

module.exports = {
  downloadHttpProtocol,
};

function isMavenCentral(pkgUrl) {
  return pkgUrl.host === 'central.maven.org';
}

function isTemporalError(err) {
  return (
    err.code === 'ECONNRESET' ||
    err.statusCode === 429 ||
    (err.statusCode > 500 && err.statusCode < 600)
  );
}

function isHostError(err) {
  return err.code === 'ETIMEDOUT';
}

function isNotFoundError(err) {
  return err.code === 'ENOTFOUND' || err.statusCode === 404;
}

function isPermissionsIssue(err) {
  return err.statusCode === 401 || err.statusCode === 403;
}

function isConnectionError(err) {
  return err.code === 'ECONNREFUSED';
}

async function downloadHttpProtocol(pkgUrl, hostType = 'maven') {
  let raw;
  try {
    raw = await got(pkgUrl, { hostType });
  } catch (err) {
    if (isNotFoundError(err)) {
      logger.debug(`Url not found ${pkgUrl}`);
    } else if (isHostError(err)) {
      // istanbul ignore next
      logger.warn({ pkgUrl }, `Cannot connect to ${hostType} host`);
    } else if (isPermissionsIssue(err)) {
      logger.warn(
        { pkgUrl },
        'Dependency lookup unauthorized. Please add authentication with a hostRule'
      );
    } else if (isTemporalError(err)) {
      logger.info({ err }, `Temporary error requesting ${pkgUrl}`);
      if (isMavenCentral(pkgUrl)) {
        throw new Error('registry-failure');
      }
    } else if (isConnectionError(err)) {
      // istanbul ignore next
      logger.info({ pkgUrl }, 'Connection refused to maven registry');
    } else {
      logger.warn({ err }, `Unknown error requesting ${pkgUrl}`);
    }
    return null;
  }
  return raw.body;
}
