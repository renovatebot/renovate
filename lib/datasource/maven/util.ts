import url from 'url';
import got from '../../util/got';
import { logger } from '../../logger';

function isMavenCentral(pkgUrl: url.URL | string): boolean {
  return (
    (typeof pkgUrl === 'string' ? pkgUrl : pkgUrl.host) === 'central.maven.org'
  );
}

function isTemporalError(err: { code: string; statusCode: number }): boolean {
  return (
    err.code === 'ECONNRESET' ||
    err.statusCode === 429 ||
    (err.statusCode >= 500 && err.statusCode < 600)
  );
}

function isHostError(err: { code: string }): boolean {
  return err.code === 'ETIMEDOUT';
}

function isNotFoundError(err: { code: string; statusCode: number }): boolean {
  return err.code === 'ENOTFOUND' || err.statusCode === 404;
}

function isPermissionsIssue(err: { statusCode: number }): boolean {
  return err.statusCode === 401 || err.statusCode === 403;
}

function isConnectionError(err: { code: string }): boolean {
  return err.code === 'ECONNREFUSED';
}

export async function downloadHttpProtocol(
  pkgUrl: url.URL | string,
  hostType = 'maven'
): Promise<string | null> {
  let raw: { body: string };
  try {
    raw = await got(pkgUrl, {
      hostType,
      hooks: {
        beforeRedirect: [
          (options: any): void => {
            if (
              options.search &&
              options.search.indexOf('X-Amz-Algorithm') !== -1
            ) {
              // maven repository is hosted on amazon, redirect url includes authentication.
              // eslint-disable-next-line no-param-reassign
              delete options.auth;
            }
          },
        ],
      },
    });
  } catch (err) {
    const failedUrl = pkgUrl.toString();
    if (isNotFoundError(err)) {
      logger.debug({ failedUrl }, `Url not found`);
    } else if (isHostError(err)) {
      // istanbul ignore next
      logger.warn({ failedUrl }, `Cannot connect to ${hostType} host`);
    } else if (isPermissionsIssue(err)) {
      logger.warn(
        { failedUrl },
        'Dependency lookup unauthorized. Please add authentication with a hostRule'
      );
    } else if (isTemporalError(err)) {
      logger.info({ failedUrl, err }, 'Temporary error');
      if (isMavenCentral(pkgUrl)) {
        throw new Error('registry-failure');
      }
    } else if (isConnectionError(err)) {
      // istanbul ignore next
      logger.info({ failedUrl }, 'Connection refused to maven registry');
    } else {
      logger.warn({ failedUrl, err }, 'Unknown error');
    }
    return null;
  }
  return raw.body;
}
