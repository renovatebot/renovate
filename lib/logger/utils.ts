import { Stream } from 'stream';

export interface BunyanRecord extends Record<string, any> {
  level: number;
  msg: string;
  module?: string;
}

const excludeProps = ['pid', 'time', 'v', 'hostname'];

export class ErrorStream extends Stream {
  private _errors: BunyanRecord[] = [];

  readable: boolean;

  writable: boolean;

  constructor() {
    super();
    this.readable = false;
    this.writable = true;
  }

  write(data: BunyanRecord) {
    const err = { ...data };
    for (const prop of excludeProps) delete err[prop];
    this._errors.push(err);
    return true;
  }

  getErrors() {
    return this._errors;
  }
}
