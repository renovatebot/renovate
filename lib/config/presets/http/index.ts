import { logger } from '../../../logger';
import { ExternalHostError } from '../../../types/errors/external-host-error';
import { Http } from '../../../util/http';
import type { HttpResponse } from '../../../util/http/types';
import { parseUrl } from '../../../util/url';
import type { Preset, PresetConfig } from '../types';
import { PRESET_DEP_NOT_FOUND, parsePreset } from '../util';

const http = new Http('preset');

export async function getPreset({
  repo: url,
}: PresetConfig): Promise<Preset | null | undefined> {
  const parsedUrl = parseUrl(url);
  let response: HttpResponse;

  if (!parsedUrl) {
    logger.debug(`Preset URL ${url} is malformed`);
    throw new Error(PRESET_DEP_NOT_FOUND);
  }

  try {
    response = await http.get(url);
  } catch (err) {
    // istanbul ignore if: not testable with nock
    if (err instanceof ExternalHostError) {
      throw err;
    }

    logger.debug(`Preset file ${url} not found`);
    throw new Error(PRESET_DEP_NOT_FOUND);
  }

  return parsePreset(response.body, parsedUrl.pathname);
}
