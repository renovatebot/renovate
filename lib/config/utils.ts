import { logger } from '../logger';
import { clone } from '../util/clone';
import * as definitions from './definitions';
import type { RenovateConfig } from './types';

export function mergeChildConfig<T, TChild>(
  parent: T,
  child: TChild
): T & TChild {
  logger.trace({ parent, child }, `mergeChildConfig`);
  if (!child) {
    return parent as never;
  }
  const parentConfig = clone(parent);
  const childConfig = clone(child);
  const config: Record<string, any> = { ...parentConfig, ...childConfig };
  for (const option of definitions.getOptions()) {
    if (
      option.mergeable &&
      childConfig[option.name] &&
      parentConfig[option.name]
    ) {
      logger.trace(`mergeable option: ${option.name}`);
      if (option.name === 'constraints') {
        config[option.name] = {
          ...parentConfig[option.name],
          ...childConfig[option.name],
        };
      } else if (option.type === 'array') {
        config[option.name] = (parentConfig[option.name] as unknown[]).concat(
          config[option.name]
        );
      } else {
        config[option.name] = mergeChildConfig(
          parentConfig[option.name] as RenovateConfig,
          childConfig[option.name] as RenovateConfig
        );
      }
      logger.trace(
        { result: config[option.name] },
        `Merged config.${option.name}`
      );
    }
  }
  return { ...config, ...config.force };
}
