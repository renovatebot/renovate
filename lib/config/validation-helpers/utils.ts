import is from '@sindresorhus/is';
import { logger } from '../../logger';
import type {
  RegexManagerConfig,
  RegexManagerTemplates,
} from '../../modules/manager/custom/regex/types';
import { regEx } from '../../util/regex';
import type { ValidationMessage } from '../types';

export function getParentName(parentPath: string | undefined): string {
  return parentPath
    ? parentPath
        .replace(regEx(/\.?encrypted$/), '')
        .replace(regEx(/\[\d+\]$/), '')
        .split('.')
        .pop()!
    : '.';
}

export function validatePlainObject(
  val: Record<string, unknown>,
): true | string {
  for (const [key, value] of Object.entries(val)) {
    if (!is.string(value)) {
      return key;
    }
  }
  return true;
}

export function validateNumber(
  key: string,
  val: unknown,
  allowsNegative: boolean,
  currentPath?: string,
  subKey?: string,
): ValidationMessage[] {
  const errors: ValidationMessage[] = [];
  const path = `${currentPath}${subKey ? '.' + subKey : ''}`;
  if (is.number(val)) {
    if (val < 0 && !allowsNegative) {
      errors.push({
        topic: 'Configuration Error',
        message: `Configuration option \`${path}\` should be a positive integer. Found negative value instead.`,
      });
    }
  } else {
    errors.push({
      topic: 'Configuration Error',
      message: `Configuration option \`${path}\` should be an integer. Found: ${JSON.stringify(
        val,
      )} (${typeof val}).`,
    });
  }

  return errors;
}

/**  An option is a false global if it has the same name as a global only option
 *   but is actually just the field of a non global option or field an children of the non global option
 *   eg. token: it's global option used as the bot's token as well and
 *   also it can be the token used for a platform inside the hostRules configuration
 */
export function isFalseGlobal(
  optionName: string,
  parentPath?: string,
): boolean {
  if (parentPath?.includes('hostRules')) {
    if (
      optionName === 'token' ||
      optionName === 'username' ||
      optionName === 'password'
    ) {
      return true;
    }
  }

  return false;
}

function hasField(
  customManager: Partial<RegexManagerConfig>,
  field: string,
): boolean {
  const templateField = `${field}Template` as keyof RegexManagerTemplates;
  return !!(
    customManager[templateField] ??
    customManager.matchStrings?.some((matchString) =>
      matchString.includes(`(?<${field}>`),
    )
  );
}

export function validateRegexManagerFields(
  customManager: Partial<RegexManagerConfig>,
  currentPath: string,
  errors: ValidationMessage[],
): void {
  if (is.nonEmptyArray(customManager.matchStrings)) {
    for (const matchString of customManager.matchStrings) {
      try {
        regEx(matchString);
      } catch (err) {
        logger.debug(
          { err },
          'customManager.matchStrings regEx validation error',
        );
        errors.push({
          topic: 'Configuration Error',
          message: `Invalid regExp for ${currentPath}: \`${matchString}\``,
        });
      }
    }
  } else {
    errors.push({
      topic: 'Configuration Error',
      message: `Each Custom Manager must contain a non-empty matchStrings array`,
    });
  }

  const mandatoryFields = ['currentValue', 'datasource'];
  for (const field of mandatoryFields) {
    if (!hasField(customManager, field)) {
      errors.push({
        topic: 'Configuration Error',
        message: `Regex Managers must contain ${field}Template configuration or regex group named ${field}`,
      });
    }
  }

  const nameFields = ['depName', 'packageName'];
  if (!nameFields.some((field) => hasField(customManager, field))) {
    errors.push({
      topic: 'Configuration Error',
      message: `Regex Managers must contain depName or packageName regex groups or templates`,
    });
  }
}
