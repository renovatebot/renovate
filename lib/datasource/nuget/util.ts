import url from 'url';
import { logger } from '../../logger';
import { regEx } from '../../util/regex';

export function parseRegistryUrl(registryUrl: string): {
  feedUrl: string;
  protocolVersion: number;
} {
  try {
    const parsedUrl = url.parse(registryUrl);
    let protocolVersion = 2;
    const protocolVersionRegExp = regEx(/#protocolVersion=(2|3)/);
    const protocolVersionMatch = protocolVersionRegExp.exec(parsedUrl.hash);
    if (protocolVersionMatch) {
      parsedUrl.hash = '';
      protocolVersion = Number.parseInt(protocolVersionMatch[1], 10);
    } else if (parsedUrl.pathname.endsWith('.json')) {
      protocolVersion = 3;
    }
    return { feedUrl: url.format(parsedUrl), protocolVersion };
  } catch (err) {
    logger.debug({ err }, `nuget registry failure: can't parse ${registryUrl}`);
    return { feedUrl: registryUrl, protocolVersion: null };
  }
}
