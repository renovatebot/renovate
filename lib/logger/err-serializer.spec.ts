import * as httpMock from '../../test/httpMock';
import { partial } from '../../test/util';
import * as hostRules from '../util/host-rules';
import { Http } from '../util/http';
import configSerializer from './err-serializer';
import { sanitizeValue } from './utils';

describe('logger/err-serializer', () => {
  it('expands errors', () => {
    const err = partial<Error & Record<string, unknown>>({
      a: 1,
      b: 2,
      message: 'some message',
      response: {
        body: 'some response body',
        url: 'some/path',
      },
      options: {
        headers: {
          authorization: 'Bearer abc',
        },
      },
    });
    expect(configSerializer(err)).toMatchSnapshot();
  });
  it('handles missing fields', () => {
    const err = partial<Error & Record<string, unknown>>({
      a: 1,
      stack: 'foo',
      body: 'some body',
    });
    expect(configSerializer(err)).toMatchSnapshot();
  });

  describe('got', () => {
    const baseUrl = 'https://github.com';

    beforeEach(() => {
      // reset module
      jest.resetAllMocks();
      httpMock.setup();
      // clean up hostRules
      hostRules.clear();
      hostRules.add({
        hostType: 'any',
        baseUrl,
        token: 'token',
      });
    });
    afterEach(() => httpMock.reset());

    it('handles http error', async () => {
      httpMock
        .scope(baseUrl)
        .post('/api')
        .reply(412, { err: { message: 'failed' } });
      let err: any;
      try {
        await new Http('any').postJson('https://:token@github.com/api');
      } catch (error) {
        err = configSerializer(error);
      }

      expect(httpMock.getTrace()).toMatchSnapshot();
      expect(err).toBeDefined();

      expect(err.response.body).toBeDefined();
      expect(err.options).toBeDefined();

      // remove platform related props
      delete err.timings;
      delete err.stack;

      // sanitize like Bunyan
      expect(sanitizeValue(err)).toMatchSnapshot();
    });
  });
});
