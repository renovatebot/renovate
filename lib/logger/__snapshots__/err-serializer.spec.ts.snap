// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger/err-serializer expands errors 1`] = `
Object {
  "a": 1,
  "b": 2,
  "message": "some message",
  "options": Object {
    "headers": Object {
      "authorization": "Bearer abc",
    },
  },
  "response": Object {
    "body": "some response body",
    "url": "some/path",
  },
}
`;

exports[`logger/err-serializer got handles http error 1`] = `
Array [
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "authorization": "Basic OnRva2Vu",
      "host": "github.com",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "POST",
    "url": "https://:token@github.com/api",
  },
]
`;

exports[`logger/err-serializer got handles http error 2`] = `
Object {
  "code": undefined,
  "message": "Response code 412 (Precondition Failed)",
  "name": "HTTPError",
  "options": Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "http2": false,
    "method": "POST",
    "password": "***********",
    "url": "https://:**redacted**@github.com/api",
    "username": "",
  },
  "response": Object {
    "body": Object {
      "err": Object {
        "message": "failed",
      },
    },
    "headers": Object {
      "content-type": "application/json",
    },
    "statusCode": 412,
    "statusMessage": "Precondition Failed",
  },
}
`;

exports[`logger/err-serializer handles missing fields 1`] = `
Object {
  "a": 1,
  "body": "some body",
  "stack": "foo",
}
`;
